### PR TITLE
examples: invoke tti example with --constant argument

### PR DESCRIPTION
**Pull request by ofmla**
I just realized that `python ~/devito/examples/seimic/tti/tti_example.py --constant` does not use the `constant-tti` model. This PR adds a very simple solution which will work for the `constant-tti `and `layers-tti` models

### CODE REVIEW

File: examples/seismic/tti/tti_example.py
Comment (FabioLuporini) [inline examples/seismic/tti/tti_example.py:43]: this shouldn't be required strictly speaking?

File: examples/seismic/tti/tti_example.py
Comment (FabioLuporini) [inline examples/seismic/tti/tti_example.py:46]: perhaps the update should be performed internally to `forward`? 

if you had multiple calls to `forward` and each time you had to update one or more parameters "by hand", it would be painful

File: examples/seismic/tti/tti_example.py
Comment (FabioLuporini) [inline examples/seismic/tti/tti_example.py:94]:
```
else:
    preset = 'layers-tii'
```

File: examples/seismic/model.py
Comment (FabioLuporini) [inline examples/seismic/model.py:378]: We need to put this under CI somehow

**Review by FabioLuporini — COMMENTED** [review]
Thanks a lot! I just left a few comments

File: examples/seismic/tti/tti_example.py
Comment (mloubout) [reply examples/seismic/tti/tti_example.py:43]: I prefer the runtime argument `vp=v0` to avoid keeping track of what was changed in the model

File: examples/seismic/tti/tti_example.py
Comment (mloubout) [reply examples/seismic/tti/tti_example.py:46]: Should probably yes be internal so the `dt` directly computed from runtime vo

File: examples/seismic/tti/tti_example.py
Comment (mloubout) [reply examples/seismic/tti/tti_example.py:94]: Don't particularly like these additional presets. There is also no reason constant would not be compatible with no azimuth

**Review by mloubout — COMMENTED** [review]
I overall don't have an issue with this as it adds test/example cases but I think it should be done as a proper constant tti as currently it just slightly modifies the current VP constant only

File: examples/seismic/tti/tti_example.py
Comment (ofmla) [reply examples/seismic/tti/tti_example.py:43]: Well, `v0` is a constant and passing it directly to `solver.forward` won't work. It gives the following error:
```
Traceback (most recent call last):
  File "devito/examples/seismic/tti/tti_example.py", line 103, in <module>
    run(shape=shape, spacing=spacing, nbl=args.nbl, tn=tn,
  File "devito/examples/seismic/tti/tti_example.py", line 48, in run
    solver.forward(vp=2., save=save)
  File "/home/oscar/devito/examples/seismic/tti/wavesolver.py", line 155, in forward
    summary = self.op_fwd(save).apply(src=src, rec=rec, u=u, v=v,
  File "/home/oscar/devito/devito/operator/operator.py", line 708, in apply
    args = self.arguments(**kwargs)
  File "/home/oscar/devito/devito/operator/operator.py", line 557, in arguments
    args = self._prepare_arguments(**kwargs)
  File "/home/oscar/devito/devito/operator/operator.py", line 449, in _prepare_arguments
    args.update(p._arg_values(**kwargs))
  File "/home/oscar/devito/devito/types/dense.py", line 843, in _arg_values
    for i, s in zip(self.dimensions, new.shape):
AttributeError: 'float' object has no attribute 'shape'
```


and for that reason I used `model.update`

File: examples/seismic/tti/tti_example.py
Comment (ofmla) [reply examples/seismic/tti/tti_example.py:46]: So, the same applies here: passing an scalar to `solver.forward() `would lead to an error
```
Traceback (most recent call last):
  File "devito/examples/seismic/tti/tti_example.py", line 103, in <module>
    run(shape=shape, spacing=spacing, nbl=args.nbl, tn=tn,
  File "devito/examples/seismic/tti/tti_example.py", line 48, in run
    solver.forward(vp=2., save=save)
  File "/home/oscar/devito/examples/seismic/tti/wavesolver.py", line 155, in forward
    summary = self.op_fwd(save).apply(src=src, rec=rec, u=u, v=v,
  File "/home/oscar/devito/devito/operator/operator.py", line 708, in apply
    args = self.arguments(**kwargs)
  File "/home/oscar/devito/devito/operator/operator.py", line 557, in arguments
    args = self._prepare_arguments(**kwargs)
  File "/home/oscar/devito/devito/operator/operator.py", line 449, in _prepare_arguments
    args.update(p._arg_values(**kwargs))
  File "/home/oscar/devito/devito/types/dense.py", line 843, in _arg_values
    for i, s in zip(self.dimensions, new.shape):
AttributeError: 'float' object has no attribute 'shape'
```
If you planned for this to be possible, then it is likely that there is some error somewhere else. If this is not the case, you need to use `model.update `before

File: examples/seismic/tti/tti_example.py
Comment (ofmla) [reply examples/seismic/tti/tti_example.py:94]: I wrote it in that way because there is not a` 'constant-noazimuth' `preset model. In any case you would have to play with conditionals I think

File: examples/seismic/model.py
Comment (ofmla) [reply examples/seismic/model.py:378]: > I overall don't have an issue with this as it adds test/example cases but I think it should be done as a proper constant tti as currently it just slightly modifies the current VP constant only

I tried to fix a mistake (constant case was not running) without changing the test design. The test only update the `vp` parameter, so I left the things as they were.

**Comment by ofmla** [comment]
> I overall don't have an issue with this as it adds test/example cases but I think it should be done as a proper constant tti as currently it just slightly modifies the current VP constant only

I tried to fix a mistake (constant case was not running) without changing the test design. The test only update the vp parameter, so I left the things as they were.

File: examples/seismic/tti/tti_example.py
Comment (mloubout) [reply examples/seismic/tti/tti_example.py:46]: This looks like it ends up in the wrong spot (dense) instead of constant.

Comment (ofmla) [reply examples/seismic/tti/tti_example.py:46]: So, are you saying you planned that the user could set a physical parameter i.e. velocity as a `Constant `or scalar  and pass it to the forward method of the `solver`, and it should work smoothly?

File: examples/seismic/tti/tti_example.py
Comment (ofmla) [reply examples/seismic/tti/tti_example.py:88]: That is what the documentation says:

ArgumentParser.error(message)[¶](https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.error)
This method prints a usage message including the message to the standard error and terminates the program with a status code of 2.

**Comment by ofmla** [comment]
> What's still missing:
> 
> * CI test for the constant case
> * (maybe) moving the "model.update" inside `solver.forward`

Well, the constant case is already tested in  `tests/test_tti.py`, I mean the forward modeling for TTI case is executed using a constant model. It seems to me that  there is not any wave-equation_example.py (i.e. `tti_example.py`, `acoustic_example.py`) in CI and that's OK. The point of this PR is that the `tti_example.py` script uses a layered model by default and although the user passes the argument `--constant` the exemple won't run on a constant model. The PR changes this.

On the other hand the `solver.forward`  gets the physical parameters from `model`  or `kwargs`. If I pass `vp` as argument to `solver.forward`  it is passed to `self.op_fwd(save).apply(...,**kwargs)` even updating `model` via `model.update` within `solver.forward`. As result I get the same errors I mentioned in the coments above. That is why I use `model.update` before. If `vp` is a numpy array there is no problem passing it to `solver.forward` but if it is a scalar or `Constant` breakage will occur.

File: examples/seismic/tti/tti_example.py
Comment (ofmla) [reply examples/seismic/tti/tti_example.py:86]: Is it worth spending a commit here? 😅

File: examples/seismic/model.py
Comment (ofmla) [reply examples/seismic/model.py:378]: All that comes into my mind is 
``` 
from examples.seismic import demo_model
import numpy as np


def test_model_updt(shape=(50, 50, 50), spacing=(20.0, 20.0, 20.0),
                    space_order=4, nbl=10):

    # Layered model (tti)
    tti_model = demo_model('layers-tti', shape=shape, spacing=spacing,
                           space_order=space_order, nbl=nbl)
    tpl1_set = set(tti_model.physical_parameters)

    # Layered model (isotropic)
    iso_model = demo_model('layers-isotropic', shape=shape, spacing=spacing,
                           space_order=space_order, nbl=nbl)
    tpl2_set = set(iso_model.physical_parameters)

    # Physical parameters in either set but not in the intersection.
    diff_phys_par = tuple(tpl1_set ^ tpl2_set)

    # Convert iso model in tti model
    for i in diff_phys_par:
        iso_model.update(i, getattr(tti_model, i).data[nbl:-nbl, nbl:-nbl, nbl:-nbl])
        assert np.array_equal(getattr(iso_model, i).data, getattr(tti_model, i).data)
```
I can add this script to the `tests` folder if you agree

File: examples/seismic/tti/tti_example.py
Comment (FabioLuporini) [inline examples/seismic/tti/tti_example.py:77]: I would maybe rather move this inside the nested `else`

```
if ...
    ...
else:
    if args.azi:
        ....
    else:
        preset = 'layers-tti'
```

File: tests/test_model_update.py
Comment (FabioLuporini) [inline tests/test_model_update.py:1]: We shouldn't test `examples` inside `tests` because in theory the examples folder shouldn't (couldn't) even be accessible.

So I've been thinking about this...

How about we add it somewhere here? https://github.com/devitocodes/devito/blob/master/examples/seismic/tti/tti_example.py#L72

It should then be captured by CI through this line? 
https://github.com/devitocodes/devito/blob/master/.github/workflows/examples.yml#L49 
(@mloubout )

File: examples/seismic/tti/tti_example.py
Comment (FabioLuporini) [inline examples/seismic/tti/tti_example.py:40]: sorry I'm confused now... why is this being entirely dropped?

File: examples/seismic/tti/tti_example.py
Comment (ofmla) [reply examples/seismic/tti/tti_example.py]: Ok

File: tests/test_model_update.py
Comment (ofmla) [reply tests/test_model_update.py]: I moved the model update test to tti_exemple.py as suggested

File: examples/seismic/tti/tti_example.py
Comment (ofmla) [reply examples/seismic/tti/tti_example.py:40]: As I see it, those lines are not longer necessary

**Review by FabioLuporini — APPROVED** [review]
LGTM. Thanks!

File: examples/seismic/tti/tti_example.py
Comment (mloubout) [inline examples/seismic/tti/tti_example.py:68]: Some nitpicking:
- Can this be moved in `model.py` 
- Can we test all cases, i.e 1) replacing the value of existing param by an array, 2) replacing existing param with a new Function 3) new param array, 4) new param Function

File: examples/seismic/tti/tti_example.py
Comment (mloubout) [inline examples/seismic/tti/tti_example.py:100]: Can probably make it a dict, but fine like that too.

**Review by mloubout — COMMENTED** [review]
Thanks for the patience and updates. Left a minor comment on the test.

I still have one main point of discussion. So the constant TTI case is not `Constant` right, it's `Function` with constant values. 
The main idea behind the constant isotropic was to have the `Constant` case to see the perf difference when there is no variation. I don't think there is much interest in a constant model if using Function personally (@FabioLuporini feel free to shit me down) since it's just a normal one with different values. So I'd rather make this an actual true constant model or remove it if there is no need for it. 

Would still keep the update to Model and test OC.

**Comment by ofmla** [comment]
> Thanks for the patience and updates. Left a minor comment on the test.
> 
> I still have one main point of discussion. So the constant TTI case is not `Constant` right, it's `Function` with constant values. The main idea behind the constant isotropic was to have the `Constant` case to see the perf difference when there is no variation. I don't think there is much interest in a constant model if using Function personally (@FabioLuporini feel free to shit me down) since it's just a normal one with different values. So I'd rather make this an actual true constant model or remove it if there is no need for it.
> 
> Would still keep the update to Model and test OC.

You are right. I reversed the changes made and kept the model updating, thx

**Comment by mloubout** [comment]
Looks like this broke some notebooks and flake8 not passing.

**Comment by ofmla** [comment]
> Looks like this broke some notebooks and flake8 not passing.

I made some changes. I hope that it will work.

File: examples/seismic/acoustic/acoustic_example.py
Comment (FabioLuporini) [reply examples/seismic/acoustic/acoustic_example.py:43]: I will merge it as is because it's been around for too long, but we should check if there's a pep8 rule for flake8 to intercept it

File: examples/seismic/tti/tti_example.py
Comment (FabioLuporini) [inline examples/seismic/tti/tti_example.py:40]: @ofmla wait... so this way it will run twice if preset=constant-tti for example ? right before this `if` and then inside?

File: examples/seismic/tti/tti_example.py
Comment (ofmla) [reply examples/seismic/tti/tti_example.py:95]: Well, it is never easy to satisfy all parties. I had written without `else` but one of you wanted an if-else. That is a matter of taste.

File: examples/seismic/acoustic/acoustic_example.py
Comment (ofmla) [reply examples/seismic/acoustic/acoustic_example.py:43]: I don't see any problem here

File: examples/seismic/tti/tti_example.py
Comment (mloubout) [reply examples/seismic/tti/tti_example.py:55]: It should be something like

```
k, v in d.items():
            d[k] = Constant(name=k, value=v, dtype=np.float32)
solver.forward(save=save, **d)
```

File: examples/seismic/acoustic/acoustic_example.py
Comment (mloubout) [reply examples/seismic/acoustic/acoustic_example.py:43]: He means there is an extra space between `With` and `a`

File: examples/seismic/tutorials/15_tti_qp_pure.ipynb
Comment (ofmla) [reply examples/seismic/tutorials/15_tti_qp_pure.ipynb:88]: I got an error in cell 6 and I just didn't want to spend much time on it. The error is the following
```
Traceback (most recent call last):
  File "tti_qpure.py", line 77, in <module>
    optime=Operator([update_p] + src_term + rec_term)
  File "/home/oscar/devito/devito/operator/operator.py", line 159, in __new__
    op = cls._build(expressions, **kwargs)
  File "/home/oscar/devito/devito/operator/operator.py", line 183, in _build
    expressions = cls._lower_exprs(expressions, **kwargs)
  File "/home/oscar/devito/devito/tools/timing.py", line 76, in __call__
    retval = self.func(*args, **kwargs)
  File "/home/oscar/devito/devito/operator/operator.py", line 283, in _lower_exprs
    expressions = flatten([i.evaluate for i in expressions])
  File "/home/oscar/devito/devito/operator/operator.py", line 283, in <listcomp>
    expressions = flatten([i.evaluate for i in expressions])
  File "/home/oscar/anaconda3/envs/devito/lib/python3.8/site-packages/cached_property.py", line 36, in __get__
    value = obj.__dict__[self.func.__name__] = self.func(obj)
  File "/home/oscar/devito/devito/types/equation.py", line 86, in evaluate
    lhs, rhs = self.lhs.evaluate, self.rhs._eval_at(self.lhs).evaluate
  File "/home/oscar/devito/devito/types/lazy.py", line 47, in evaluate
    args = self._evaluate_args()
  File "/home/oscar/devito/devito/types/lazy.py", line 42, in _evaluate_args
    return [Evaluable._evaluate_maybe_nested(i) for i in self.args]
  File "/home/oscar/devito/devito/types/lazy.py", line 42, in <listcomp>
    return [Evaluable._evaluate_maybe_nested(i) for i in self.args]
  File "/home/oscar/devito/devito/types/lazy.py", line 16, in _evaluate_maybe_nested
    return maybe_evaluable.evaluate
  File "/home/oscar/devito/devito/finite_differences/derivative.py", line 322, in evaluate
    return self._eval_fd(self.expr)
  File "/home/oscar/devito/devito/finite_differences/derivative.py", line 356, in _eval_fd
    res = generic_derivative(expr, *self.dims, self.fd_order, self.deriv_order,
  File "/home/oscar/devito/devito/finite_differences/tools.py", line 50, in wrapper
    return S.Zero if expr.is_Number else func(expr, *args, **kwargs)
  File "/home/oscar/devito/devito/finite_differences/tools.py", line 67, in wrapper
    return func(expr, *args, **kwargs)
  File "/home/oscar/devito/devito/finite_differences/finite_difference.py", line 195, in generic_derivative
    return make_derivative(expr, dim, fd_order, deriv_order, side, matvec, x0, symbolic)
  File "/home/oscar/devito/devito/finite_differences/finite_difference.py", line 208, in make_derivative
    return indices_weights_to_fd(expr, dim, indices, weights, matvec=matvec.val)
  File "/home/oscar/devito/devito/finite_differences/finite_difference.py", line 223, in indices_weights_to_fd
    iloc -= expr.indices_ref[dim] - dim
  File "/home/oscar/devito/devito/tools/data_structures.py", line 48, in __getitem__
    return self.__getitem_hook__(key)
  File "/home/oscar/devito/devito/types/utils.py", line 27, in __getitem_hook__
    raise KeyError
KeyError

``` 
I can open an issue with a MFE after you merge the PR, it is the easier way. What do you think

File: examples/seismic/tti/tti_example.py
Comment (ofmla) [reply examples/seismic/tti/tti_example.py]: ok

File: examples/seismic/tutorials/15_tti_qp_pure.ipynb
Comment (mloubout) [reply examples/seismic/tutorials/15_tti_qp_pure.ipynb:88]: Ok that's fine.

**Review by mloubout — APPROVED** [review]

**Comment by mloubout** [comment]
Needs rebase and GTG

**Comment by ofmla** [comment]
> Needs flake8, rebase and GTG

The flake8 errors are in files that I didn't touch. It is weird

**Comment by mloubout** [comment]
>The flake8 errors are in files that I didn't touch. It is weird

Must be the new standards, is fixed in #1979 so I'd say wait for it to be merged then rebase and should be all good

File: examples/seismic/tti/tti_example.py
Comment (FabioLuporini) [inline examples/seismic/tti/tti_example.py:54]: shuold be a dict comprehension

**Pull request merged**
